### ISSUE TITLE
Storage: [triggerFunction] does not exist in template, removed line but requires a value

### ISSUE THREAD

**Issue by aflatness** [aws-amplify/amplify-cli#8948]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.

### How did you install the Amplify CLI?

curl

### If applicable, what version of Node.js are you using?

14.16.1

### Amplify CLI Version

7.4.1

### What operating system are you using?

Mac

### Amplify Categories

storage

### Amplify Commands

push

### Describe the bug

When trying to make changes to the API category, the stack is failing to build the storage stack:

`UPDATE_FAILED      xxx       AWS::CloudFormation::Stack Thu Nov 18 2021 13:58:05 GMT-0700 Parameters: [triggerFunction] do not exist in the template`

Initial storage was created through cli, and was built with the same .json files. Not currently using any triggerFunction, and I have it set to "NONE".
I didn't make any changes to the storage since last build so unsure why this would be failing now.

### Expected behavior

When running `amplify push`, the Functions and API categories should be the only ones updated, but is failing to due the storage stack failing to build. The storage stack should build fine, as I haven't made any changes to the stack.

### Reproduction steps

1. Create storage through cli without a trigger function and build
2. Make change to another stack
3. Storage stack fails to build due to requiring a triggerFunction

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
# Put your logs below this line

UPDATE_FAILED      xxx       AWS::CloudFormation::Stack Thu Nov 18 2021 13:58:05 GMT-0700 Parameters: [triggerFunction] do not exist in the template 


```

</details>

**Comment by sachscode** [comment]
@aflatness 
1. Could you please share the sequence of commands which led you to this state. I will try to repro the issue on my mac.
2. Could you please share the contents of your amplify folder . 
3. Was this a new project ? or are you operating on an existing project.

**Comment by kaustavghosh06** [comment]
also, this might be helpful - https://github.com/aws-amplify/amplify-cli/issues/8892#issuecomment-970933361 to unblock.

**Comment by aflatness** [comment]
@kaustavghosh06 Appreciate the link. I did try removing both those lines, but then a new failure appears for `UpdateRolesWithIDPFunctionRole`.
I was able to determine that in cloudFormation, there was no reference to triggerFunction in the template or parameters for the storage stack. But we had it referenced for both in our local files.

@sachscode 
1. Update local API schema unrelated to storage. Run `amplify push`
2.
cli.json
```
{
  "features": {
    "graphqltransformer": {
      "addmissingownerfields": true,
      "improvepluralization": true,
      "validatetypenamereservedwords": true,
      "useexperimentalpipelinedtransformer": false,
      "enableiterativegsiupdates": true,
      "secondarykeyasgsi": true,
      "skipoverridemutationinputtypes": true
    },
    "frontend-ios": {
      "enablexcodeintegration": true
    },
    "auth": {
      "enablecaseinsensitivity": true,
      "useinclusiveterminology": true,
      "breakcirculardependency": true
    },
    "codegen": {
      "useappsyncmodelgenplugin": true,
      "usedocsgeneratorplugin": true,
      "usetypesgeneratorplugin": true,
      "cleangeneratedmodelsdirectory": true,
      "retaincasestyle": true,
      "addtimestampfields": true,
      "handlelistnullabilitytransparently": true,
      "emitauthprovider": true,
      "generateindexrules": true,
      "enabledartnullsafety": true
    },
    "appsync": {
      "generategraphqlpermissions": true
    }
  }
}
```
3. This is an existing project. Storage was initiated through the cli about 2 weeks ago and has been fine since. Today I tried pushing an update not related to storage, and the storage stack fails to build.

**Comment by kaustavghosh06** [comment]
@aflatness what's the error you're getting with `UpdateRolesWithIDPFunctionRole`?

**Comment by aflatness** [comment]
> @aflatness what's the error you're getting with `UpdateRolesWithIDPFunctionRole`?

`Resource update cancelled`. Also appears for our storage stack.
We do have another issue regarding other parameters for our Auth stack, which also hasn't been updated, where it's now requiring values for `breakCircularDependency` and `useEnabledMfas`.

**Comment by kaustavghosh06** [comment]
Getting these two things would be helpful
1. Stack trace/event-trace of the  failure
2. The contents of the amplify/ folder - you can send us a sharable link to amplify-cli@amazon.com to investigate.

**Comment by aflatness** [comment]
@kaustavghosh06 
```
⠧ Updating resources in the cloud. This may take a few minutes...

UPDATE_IN_PROGRESS xxMainStackxx AWS::CloudFormation::Stack Thu Nov 18 2021 15:47:34 GMT-0700 (Mountain Standard Time) User Initiated                                            
UPDATE_IN_PROGRESS xxStorageStackxx          AWS::CloudFormation::Stack Thu Nov 18 2021 15:47:38 GMT-0700 (Mountain Standard Time)                                                           
UPDATE_FAILED      xxStorageStackxx                      AWS::CloudFormation::Stack Thu Nov 18 2021 15:47:39 GMT-0700 (Mountain Standard Time) Parameters: [triggerFunction] do not exist in the template
UPDATE_IN_PROGRESS UpdateRolesWithIDPFunctionRole  AWS::IAM::Role             Thu Nov 18 2021 15:47:39 GMT-0700 (Mountain Standard Time)                                                     ⠋ Updating resources in the cloud. This may take a few minutes...

UPDATE_IN_PROGRESS xxAuthStackxx               AWS::CloudFormation::Stack Thu Nov 18 2021 15:47:39 GMT-0700 (Mountain Standard Time)                                                                       
UPDATE_FAILED      UpdateRolesWithIDPFunctionRole AWS::IAM::Role             Thu Nov 18 2021 15:47:39 GMT-0700 (Mountain Standard Time) Resource update cancelled                                             
UPDATE_FAILED      xxAuthStackxx               AWS::CloudFormation::Stack Thu Nov 18 2021 15:47:40 GMT-0700 (Mountain Standard Time) Parameters: [breakCircularDependency, useEnabledMfas]⠹ Updating resources in the cloud. This may take a few minutes...

UPDATE_ROLLBACK_IN_PROGRESS xxMainStackxx AWS::CloudFormation::Stack Thu Nov 18 2021 15:47:40 GMT-0700 (Mountain Standard Time) The following resource(s) failed to update: [xxMainStackxx, xxStorageStackxx, UpdateRolesWithIDPFunctionRole]⠹ Updating resources in the cloud. This may take a few minutes...

UPDATE_COMPLETE UpdateRolesWithIDPFunctionRole AWS::IAM::Role             Thu Nov 18 2021 15:48:07 GMT-0700 (Mountain Standard Time) 
UPDATE_COMPLETE xxAuthStackxx               AWS::CloudFormation::Stack Thu Nov 18 2021 15:48:07 GMT-0700 (Mountain Standard Time) 
UPDATE_COMPLETE xxStorageStackxx                     AWS::CloudFormation::Stack Thu Nov 18 2021 15:48:07 GMT-0700 (Mountain Standard⠼ Updating resources in the cloud. This may take a few minutes...

UPDATE_ROLLBACK_COMPLETE_CLEANUP_IN_PROGRESS xxMainStackxx AWS::CloudFormation::Stack Thu Nov 18 2021 15:48:⠹ Updating resources in the cloud. This may take a few minutes...

UPDATE_COMPLETE          xxStorageStackxx                      AWS::CloudFormation::Stack Thu Nov 18 2021 15:48:11 GMT-0700 (Mountain Standard Time) 
UPDATE_COMPLETE          xxAuthStackxx                AWS::CloudFormation::Stack Thu Nov 18 2021 15:48:11 GMT-0700 (Mountain Standard Time) 
UPDATE_ROLLBACK_COMPLETE xxMainStackxx AWS::CloudFormation::Stack Thu Nov 18 2021 15:48:11 GMT-0700 (Mountai⠙ Updating resources in the cloud. This may take a few minutes...

Following resources failed

Resource Name: xxMainStackxx-authRole-idp (AWS::IAM::Role)
Event Type: update
Reason: Resource update cancelled
URL: 

✖ An error occurred when pushing the resources to the cloud
🛑 An error occurred during the push operation: Resource is not in the state stackUpdateComplete
```

**Comment by kaustavghosh06** [comment]
Yeah, I'm curious to know about the error and stack trace post removing the triggerFunction to investigate further.

**Comment by n2westman** [comment]
https://github.com/aws-amplify/amplify-cli/issues/8917

I'm pretty sure this is the same issue (or at least related). I don't think @aflatness stressed it enough, but we made no changes to the backend, and deploys started failing. Other issues masked it, but I'm now seeing effectively the same thing.

**Comment by n2westman** [comment]
So pretty much the state we're in is the following `amplify env pull --restore` then `amplify push --force` yields the following error (I have _zero_ clue why the CLI is trying to update the user pool):

```UPDATE_FAILED      UserPool      AWS::Cognito::UserPool Thu Nov 18 2021 21:16:21 GMT-0800 (Pacific Standard Time) Invalid AttributeDataType input, consider using the provided AttributeDataType enum. (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException; Request ID: a0d62e43-5c82-4161-b99f-9943eabfb0fc; Proxy: null)
UPDATE_FAILED      MFALambdaRole AWS::IAM::Role         Thu Nov 18 2021 21:16:22 GMT-0800 (Pacific Standard Time) Resource update cancelled    ```

**Comment by n2westman** [comment]
Update - adding running `amplify update auth` seems to have fixed it. It blew away any inconsistent state and re-wrote the auth configs to work again.

Regarding the `triggerFunction` - this diff seemed to fix it.
![Screen Shot 2021-11-19 at 10 09 33 AM](https://user-images.githubusercontent.com/1989023/142670919-1f29422c-8aa4-444a-91c9-acc1d3f11000.png)

**Comment by kaustavghosh06** [comment]
@n2westman thanks for the pointers and glad you were able to get back on track! 
@aflatness are you still stuck on this issue? Please let us know and we'd re-open the issue.
